### DISCUSSION TITLE
자료구조 - 스택/큐/덱 - 은진&연서

### DISCUSSION THREAD

**Discussion by watty-exclamationMark** [Shortudy-10th/tech-for-developer#49] [질의 응답]
+ 큐, 스택, 덱이 각각 어떠한 구조를 가지는지 설명해 주세요.
+ 스택, 큐, 덱을 프로그램상에서 구현한다면, 어떻게 구현할 수 있을까요?
+ OS, 어플리케이션 등 다양한 프로그램에서 스택, 큐, 덱을 실제로 어떻게 활용하고 있나요? 그 예시에 대해 설명해주세요.
+ 브라우저에서 스택과 큐는 언제 활용되고 있나요?
+ 스택과 큐의 특징과 이 둘은 어디에 주로 사용되나요?

**Comment by watty-exclamationMark** [comment]
### ❓️큐, 스택, 덱이 각각 어떠한 구조를 가지는지 설명해 주세요.

### 💬답변
> 1. 큐
> 
> ![image](https://github.com/Shortudy-10th/tech-for-developer/assets/70586307/0392701a-8ee5-4f82-9bf1-2e2799959e7e)
> (너무 조잡해...)
> FIFO(First In First Out)의 구조를 가지고 있다. 단방향 자료구조.
>
> ---
>
> 2. 스택
>
> ![image](https://github.com/Shortudy-10th/tech-for-developer/assets/70586307/6f3322e2-4855-4193-9031-bee45a0a09bd)
> (디자인의 한계...)
> LIFO(Last In First Out)의 구조를 가지고 있다. 입구와 출구가 같은 구조.
>
> ---
>
> 3. 덱
>
> ![image](https://github.com/Shortudy-10th/tech-for-developer/assets/70586307/5986633c-c9d2-4ccb-88ed-d9c12ee3f543)
>
> 
> 양방향 자료구조. 어느 쪽이든 in / out을 할 수 있다.
> 
>         
> ---

**Comment by watty-exclamationMark** [comment]
### ❓️스택, 큐, 덱을 프로그램상에서 구현한다면, 어떻게 구현할 수 있을까요?

### 💬답변

**Reply by watty-exclamationMark** [reply]
이거 뭐 코드를 써야하는건가...?

**Comment by watty-exclamationMark** [comment]
### ❓️OS, 어플리케이션 등 다양한 프로그램에서 스택, 큐, 덱을 실제로 어떻게 활용하고 있나요? 그 예시에 대해 설명해주세요. 
### 💬답변
> 1. 스택 : 함수 호출
> 프로그래밍 언어는 호출된 함수 호출을 관리할 때 스택을 사용한다. 함수가 호출될 때마다 스택에 추가되고, 함수가 종료될 때 스택에서 제거된다. 
> 덕분에 재귀 함수 호출과 중첩 함수 호출도 가능하다.
> 
> ![image](https://github.com/Shortudy-10th/tech-for-developer/assets/39663810/00b77600-ca6a-47b8-8468-90fc4205a022)
> 
> 
> 2. 큐 : 작업 스케줄링과 프로세스 관리
> 운영체제에서 다수의 프로세스를 관리하기 위해 큐를 사용한다. 대기 중인 프로세스를 큐에 넣고 순서대로 처리한다. (물론 실제로는 훨씬 더 복잡하지만...)
> 덕분에 시스템의 자원을 효율적으로 할당할 수 있다. 
> ![image](https://github.com/Shortudy-10th/tech-for-developer/assets/39663810/8bbcc2bf-17d3-4b28-aded-7c4e7adaa207)
>
> 3. 덱 : 사용자 정의 시스템
> OS는 주로 스택, 큐, 힙과 같은 더 기본적인 자료구조를 사용하여 프로세스 관리, 스케줄링, 메모리 관리 등의 작업을 수행한다.
> 따라서 사용자가 양방향으로 처리하고자 하는 다양한 로직에 활용하면 된다.
> ex) 스크롤 기능 : 화면에 표시하고 싶은 객체나 데이터를 덱에 저장하여, 스크롤 이벤트에 따라 양방향으로 객체를 추가/삭제한다.
>

**Comment by watty-exclamationMark** [comment]
### ❓️브라우저에서 스택과 큐는 언제 활용되고 있나요?
### 💬답변
1. 스택 : 웹 브라우저의 뒤로 가기 기능
웹페이지 방문 히스토리를 스택에 저장하여 구현할 수 있다. 사용자가 뒤로 가기를 클릭하면 스택의 가장 위에 있는 웹페이지로 이동하면 된다.

2. 큐 : 웹 서버 요청 처리
웹 서버는 동시에 수많은 클라이언트 요청을 받아야 하는데, 큐를 활용하여 클라이언트 요청을 순서대로 처리할 수 있다.

**Comment by watty-exclamationMark** [comment]
### ❓️스택과 큐의 특징과 이 둘은 어디에 주로 사용되나요?

### 💬답변

**Reply by aeroej** [reply]
중복이라 패스
